contracts/riskon/UniswapV3RiskOnVault.sol

```diff
@@ -0,0 +1,673 @@
+        _redeemBalance = redeemToVault(_redeemShares, _totalShares);
+        emit RedeemToVault(_redeemBalance);
+    }
+
+    /// @notice Redeem to vault
+    /// @param _redeemShares The amount of shares to withdraw
+    /// @param _totalShares The total amount of shares owned by this strategy
+    /// @return _redeemBalance The balance of redeem
+    function redeemToVault(uint256 _redeemShares, uint256 _totalShares) internal whenNotEmergency nonReentrant returns (uint256 _redeemBalance) {
+        uint256 currentBorrow = uniswapV3RiskOnHelper.getCurrentBorrow(borrowToken, interestRateMode, address(this));
+        (uint256 _totalCollateral, , , , ,) = uniswapV3RiskOnHelper.borrowInfo(address(this));
+        address aCollateralToken = uniswapV3RiskOnHelper.getAToken(wantToken);
+        if (_redeemShares == _totalShares) {
+            harvest();
+            withdraw(baseMintInfo.tokenId, _redeemShares, _totalShares);
+            withdraw(limitMintInfo.tokenId, _redeemShares, _totalShares);
+            delete baseMintInfo;
+            delete limitMintInfo;
+            uint256 borrowTokenBalance = balanceOfToken(borrowToken);
+            if (currentBorrow > borrowTokenBalance) {
+                IERC20Upgradeable(wantToken).safeApprove(RiskOnConstant.UNISWAP_V3_ROUTER, 0);
+                IERC20Upgradeable(wantToken).safeApprove(RiskOnConstant.UNISWAP_V3_ROUTER, balanceOfToken(wantToken));
+                IUniswapV3(RiskOnConstant.UNISWAP_V3_ROUTER).exactOutputSingle(IUniswapV3.ExactOutputSingleParams(wantToken, borrowToken, fee, address(this), block.timestamp, currentBorrow - borrowTokenBalance, type(uint256).max, 0));
+            }
+            __repay(currentBorrow);
+            __removeCollateral(aCollateralToken, IERC20(aCollateralToken).balanceOf(address(this)));
+        } else {
+            uint256 beforeWantTokenBalance = balanceOfToken(wantToken);
+            uint256 beforeBorrowTokenBalance = balanceOfToken(borrowToken);
+            withdraw(baseMintInfo.tokenId, _redeemShares, _totalShares);
+            withdraw(limitMintInfo.tokenId, _redeemShares, _totalShares);
+            //            uint256 redeemVaultWantTokenBalance = beforeWantTokenBalance * _redeemShares / _totalShares;
+            //            uint256 redeemVaultBorrowTokenBalance = beforeBorrowTokenBalance * _redeemShares / _totalShares;
+            //            uint256 redeemWantTokenBalance = afterWantTokenBalance - beforeWantTokenBalance;
+            uint256 redeemBorrowTokenBalance = balanceOfToken(borrowToken) - beforeBorrowTokenBalance;
+            uint256 redeemCurrentBorrow = currentBorrow * _redeemShares / _totalShares;
+            if (redeemCurrentBorrow > redeemBorrowTokenBalance) {
+                IERC20Upgradeable(wantToken).safeApprove(RiskOnConstant.UNISWAP_V3_ROUTER, 0);
+                IERC20Upgradeable(wantToken).safeApprove(RiskOnConstant.UNISWAP_V3_ROUTER, balanceOfToken(wantToken));
```

Comment:
can put L336 and 337 into function initialize,approve type(uint256).max amount for gas save
```
 IERC20Upgradeable(wantToken).safeApprove(RiskOnConstant.UNISWAP_V3_ROUTER, type(uint256).max);
```